server/api/image.py

```diff
@@ -94,6 +98,84 @@ def find(self, params):
         ]
 
     @describeRoute(
+        Description('Download multiple images as a ZIP file.')
+        .param('datasetId', 'The ID of the dataset to download.',
+               required=False)
+        .param('filter', 'Filter the images by a PegJS-specified grammar '
+                         '(causing "datasetId" to be ignored).',
+               required=False)
+        .errorResponse()
+    )
+    @access.public
+    def downloadMultiple(self, params):
+        Dataset = self.model('dataset', 'isic_archive')
+        File = self.model('file')
+        Image = self.model('image', 'isic_archive')
+
+        if 'filter' in params:
+            query = querylang.astToMongo(json.loads(params['filter']))
+        else:
+            query = {}
+            if 'datasetId' in params:
+                try:
+                    query.update({'folderId': ObjectId(params['datasetId'])})
+                except InvalidId:
+                    raise RestException(
+                        'Invalid "folderId" ObjectId: %s' % params['datasetId'])
+
+        user = self.getCurrentUser()
+        downloadFileName = 'ISIC-images'
+
+        def stream():
+            datasetCache = dict()
+            zipGenerator = ziputil.ZipGenerator(downloadFileName)
+
+            for image in Image.filterResultsByPermission(
+                    # TODO: exclude additional fields from the cursor
+                    Image.find(query, sort=[('name', 1)], fields={'meta': 0}),
+                    user=user, level=AccessType.READ, limit=0, offset=0):
+                datasetId = image['folderId']
+                if datasetId not in datasetCache:
+                    datasetCache[datasetId] = Dataset.load(
+                        datasetId, force=True, exc=True)
+                dataset = datasetCache[datasetId]
+                imageFile = Image.originalFile(image)
+                imageFileGenerator = File.download(imageFile, headers=False)
+                for data in zipGenerator.addFile(
+                        imageFileGenerator,
+                        path=os.path.join(dataset['name'], imageFile['name'])):
+                    yield data
+
+            for dataset in six.viewvalues(datasetCache):
+                licenseText = mail_utils.renderTemplate(
+                    'license_%s.mako' % dataset['meta']['license'])
```

Comment:
I hit an error here:
```
CompileException: Unicode decode operation of encoding 'ascii' failed in file '/home/vagrant/girder/plugins/isic_archive/server/license_templates/license_CC-BY.mako' at line: 0 char: 
```
This likely affects CC-BY and CC-BY-NC, since they are not ASCII:
```
attribution_CC-0.mako:     ASCII text, with CRLF line terminators
attribution_CC-BY.mako:    ASCII text, with CRLF line terminators
attribution_CC-BY-NC.mako: ASCII text, with CRLF line terminators
license_CC-0.mako:         ASCII text, with CRLF line terminators
license_CC-BY.mako:        UTF-8 Unicode text, with CRLF line terminators
license_CC-BY-NC.mako:     UTF-8 Unicode text, with CRLF line terminators
```